### ISSUE TITLE
Chromakey Video Handling Error 

### ISSUE THREAD

**Issue by suryanadiminti99** [ThomasGorisse/sceneform-android-sdk#21]
If i use chromakey video  like below , getting this error . Please look on it also . can we play videos in ARscene .  @ThomasGorisse 

**My Code :** 

`ExternalTexture texture = new ExternalTexture();
        // Create an Android MediaPlayer to capture the video on the external texture's surface.
        mediaPlayer = MediaPlayer.create(this, Uri.fromFile(new File(videoUrl)));
        mediaPlayer.setSurface(texture.getSurface());
        mediaPlayer.setLooping(false);

ModelRenderable.builder()
                .setSource(this, R.raw.chroma_key_video)
                .build()
                .thenAccept(
                        renderable -> {

                            renderable.getMaterial().setExternalTexture("videoTexture", texture);
                            renderable.getMaterial().setFloat4("keyColor", CHROMA_KEY_COLOR);

                            AnchorNode anchorNode = new AnchorNode(anchor);

                            // Create a node to render the video and add it to the anchor.
                            Node videoNode = new Node();
                            anchorNode.setParent(arFragment.getArSceneView().getScene());
                            videoNode.setParent(anchorNode);

                            // Vector3 vectorScale = new Vector3(Float.parseFloat(scalex), 0.0001f, Float.parseFloat(scaley));

                            Vector3 vectorPos = new Vector3(Float.parseFloat(posx), Float.parseFloat(posy),(Float.parseFloat(posz)));

                            Quaternion rx = Quaternion.eulerAngles(new Vector3(Float.parseFloat(rotatex), 0.0f, 0.0f));
                            Quaternion ry = Quaternion.eulerAngles(new Vector3(0.0f, Float.parseFloat(rotatey), Float.parseFloat(rotatez)));

                            // Set Scaling
                            //  videoNode.setLocalScale(vectorScale);
                            // Set Position
                            videoNode.setLocalPosition(vectorPos);
                            // Set Rotation
                            videoNode.setLocalRotation(Quaternion.multiply(rx,ry));

                            // Start playing the video when the first node is placed.
                            if (!mediaPlayer.isPlaying()) {
                                mediaPlayer.start();

                                texture.getSurfaceTexture().setOnFrameAvailableListener(new SurfaceTexture.OnFrameAvailableListener() {
                                    @Override
                                    public void onFrameAvailable(SurfaceTexture surfaceTexture) {
                                        videoNode.setRenderable(renderable);
                                        texture.getSurfaceTexture().setOnFrameAvailableListener(null);
                                    }
                                });

                            } else {
                                videoNode.setRenderable(renderable);
                            }

                        })
                .exceptionally(
                        throwable -> {
                            Toast toast =
                                    Toast.makeText(this, "Unable to load video renderable", Toast.LENGTH_LONG);
                            toast.setGravity(Gravity.CENTER, 0, 0);
                            toast.show();
                            return null;
                        });`

**Error :** 

`E/Texture: Unable to load Texture registryId='file:///storage/emulated/0/Android/data/com.plugxr.cloudanchors/files/LC4V39/BURGERKING/Assets/ua-d107746686e9acb61970044cd7c7f5f1/Content/videos/1615873117.mp4'
    java.util.concurrent.CompletionException: java.lang.IllegalStateException: Failed to decode the texture bitmap. The InputStream was not a valid bitmap.
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:276)
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:282)
        at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1627)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: java.lang.IllegalStateException: Failed to decode the texture bitmap. The InputStream was not a valid bitmap.
        at com.google.ar.sceneform.rendering.Texture$Builder.lambda$makeBitmap$1(Texture.java:334)`

**Comment by ThomasGorisse** [comment]
Did you call `mediaPlayer.prepareAsync()` before starting the video?
And why waiting for a FrameAvailable to set the node renderable ?

**Comment by suryanadiminti99** [comment]
No, I did not use mediaplayer.prepareAsync();

If frame is not available , empty node will visible . That's why i did not show up renderable if frame is not available. 

And also this reference taken from Google 1.15 scenefrom version . 

[](https://github.com/google-ar/sceneform-android-sdk/tree/v1.15.0/samples/chromakeyvideo)

**Comment by suryanadiminti99** [comment]
No, I did not use mediaplayer.prepareAsync();

If frame is not available , empty node will visible . That's why i did not show up renderable if frame is not available. 

And also this reference taken from Google 1.15 scenefrom version . 

[](https://github.com/google-ar/sceneform-android-sdk/tree/v1.15.0/samples/chromakeyvideo)

**Comment by suryanadiminti99** [comment]
I think SFB is not supported in this version. am I correct?

**Comment by ThomasGorisse** [comment]
That's right, you have to create a glTF plane with a 100% emissive material.
If you are using Blender, just use a Principal BSDF material and Filament will handle it.
Someone here already done that, please Google it.

**Comment by suryanadiminti99** [comment]
oh, Thankyou @ThomasGorisse For your quick response.

**Comment by suryanadiminti99** [comment]
Someone, please help me. I did not find the issue. If you know how to play video using glb or gltf plane.

**Comment by suryanadiminti99** [comment]
> That's right, you have to create a glTF plane with a 100% emissive material.
> If you are using Blender, just use a Principal BSDF material and Filament will handle it.
> Someone here already done that, please Google it.

@ThomasGorisse HI, Will you please share that link if possible . I cannot find that link.
I created glb renderable model . If i added external texture to that model , it will through error.

**Comment by ThomasGorisse** [comment]
Fixed here: 5c2f6ce24e2cc88490f065e53cf78eada8f0cc7a

**Comment by suryanadiminti99** [comment]
@ThomasGorisse Why you are closed this issue . Is it resolved ? Will you please explain it once.

**Comment by ThomasGorisse** [comment]
Yes. You can now set the Texture of a glTF node without issue.
Have a look at the Texture sample.

**Comment by suryanadiminti99** [comment]
> Yes. You can now set the Texture of a glTF node without issue.
> Have a look at the Texture sample.

@ThomasGorisse oh great . Thankyou .

**Comment by suryanadiminti99** [comment]
> Yes. You can now set the Texture of a glTF node without issue.
> Have a look at the Texture sample.

@ThomasGorisse 
Glb with texture is working fine .

But i applied .mp4 file to same glb . getting error like below .

`2021-03-25 11:19:57.970 13442-13512/com.plugxr.cloudanchors E/Texture: Unable to load Texture registryId='texture/rrr.mp4'
    java.util.concurrent.CompletionException: java.lang.IllegalStateException: Failed to decode the texture bitmap. The InputStream was not a valid bitmap.
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:276)
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:282)
        at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1627)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: java.lang.IllegalStateException: Failed to decode the texture bitmap. The InputStream was not a valid bitmap.
        at com.google.ar.sceneform.rendering.Texture$Builder.lambda$makeBitmap$1(Texture.java:334)
        at com.google.ar.sceneform.rendering.-$$Lambda$Texture$Builder$kapChZv8zsxavp7lXl8hPG00GJo.get(Unknown Source:4)
        at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1625)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:764) 
`
